src/composables/use-fetch-state.ts

```diff
@@ -0,0 +1,111 @@
+ * on the first app load, or after a change in search filters or search term.
+ * - `fetching`: the request was sent, but no response was received yet.
+ * - `success`: a successful response was received for the current request.
+ * - `finished`: for multi-page requests, this is true when no more pages are left.
+ * - `error`: an error response was received.
+ */
+const statuses = {
+  IDLE: 'idle',
+  FETCHING: 'fetching',
+  SUCCESS: 'success',
+  ERROR: 'error',
+  FINISHED: 'finished',
+} as const
+
+type Status = typeof statuses[keyof typeof statuses]
+/**
+ * When one of these statuses is received, the fetchError is set to `null` in a watcher.
+ */
+const nonErrorStatuses: Status[] = [
+  statuses.IDLE,
+  statuses.FETCHING,
+  statuses.SUCCESS,
+]
+/**
+ * With these statuses, it is possible to send the same request for a new page.
+ * This can help debounce requests and prevent racing states.
+ */
+const canFetchAgainStatuses: Status[] = [statuses.IDLE, statuses.SUCCESS]
+
+/* Composable */
+
+export const useFetchState = (initialState = statuses.IDLE) => {
+  const fetchStatus: Ref<Status> = ref(initialState)
+  const fetchError: Ref<string | null> = ref(null)
+
+  watch(fetchStatus, () => {
+    if (nonErrorStatuses.includes(fetchStatus.value)) {
+      fetchError.value = null
+    }
+  })
+  const reset = () => {
+    fetchStatus.value = statuses.IDLE
+  }
+  const startFetching = () => {
+    fetchStatus.value = statuses.FETCHING
+  }
+  /**
+   * Called when any response is received. For an error response,
+   * pass the errorMessage string as a parameter.
+   * @param errorMessage - the message to show for response error.
+   */
+  const endFetching = ({
+    errorMessage = undefined,
+  }: { errorMessage?: string } = {}) => {
+    if (errorMessage) {
+      fetchStatus.value = statuses.ERROR
+      fetchError.value = errorMessage
+    } else {
+      fetchStatus.value = statuses.SUCCESS
+    }
+  }
+  const setFinished = () => {
+    fetchStatus.value = statuses.FINISHED
+  }
+
+  /**
+   * Whether the current request has been sent. This can be useful for displaying
+   * initial search page with a blank search term, such as for a provider search page.
+   */
+  const hasStarted = computed(() => fetchStatus.value !== statuses.IDLE)
+  const isFetching = computed(() => fetchStatus.value === statuses.FETCHING)
+  const canFetchAgain = computed(() =>
+    canFetchAgainStatuses.includes(fetchStatus.value)
+  )
+  const isFinished = computed(() => fetchStatus.value === statuses.FINISHED)
+  const isError = computed(() => fetchStatus.value === statuses.ERROR)
+  const fetchingError = computed(() => fetchError.value)
+
+  const fetchingState: FetchState = reactive({
+    hasStarted,
+    isFetching,
+    canFetchAgain,
+    fetchingError,
+    isFinished,
+  })
+
+  return {
+    isFetching,
+    fetchingError,
+    canFetchAgain,
+    hasStarted,
+    isFinished,
+    isError,
+
+    fetchingState,
```

Comment:
What's the reason for the `fetchingState` reactive when all the data included in it is already present on the top level of the store's API?

Reply:
I meant to ask for everyone's opinion on what the best option is: 
- add `fetchState` as a sub-item to be used in components and the store as it is used now,
or
- to have everything in the top level. 
I guess having it in the top level should be enough, don't you think?

Reply:
I think having it all on the top level makes sense too. Otherwise watching any of them requires watching either the full state object _or_ using `toRefs`, which feels cumbersome.

Reply:
Ok, I removed the `fetchingState`, and then I realized that using all those single pieces separately as computed means that we would have to add `.value` to each of them every time we consume them, which isn't very user-friendly. Do you think having it as a reactive would really affect the performance, @sarayourfriend ?